Building Default Inference for Device Type

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69802
* #70209
* #69051
* #69050
* #71184
* #69052
* __->__ #69049

Differential Revision: [D33555885](https://our.internmc.facebook.com/intern/diff/D33555885)

Gamrix committed:
Building Default Inference for Device Type

[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

eellison reviewed (commented):
can we just unify this with dtype analysis ? there isnt much of a reason to do one without the other

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+  // Detecting of the op has a device object argument
+  // as there is implicit string conversion to device
+  auto arguments = n->getOperator().schema().arguments();
+  auto input_vec = n->inputs();
```
Do we need to handle the zero-dim case here?  `torch.rand([4, 4]).cuda() + torch.tensor(0.1)` ? The device is not taken from inputs

Gamrix replied:
This would be part of the `propWithNoDevice` function, but ugh, I didn't realize that there was that edge case.

eellison replied:
I think we can update the heuristics to something along the line of the following
a) take the device of a tensor if it has size set and it's dim > 0
b) if all tensors have the same device, take it

eellison replied:
I think we also need to exclude list :
reshape_as,
view_as,
expand_as

Gamrix replied:
Currently they should be properly handled by the logic of "return no device if there are multiple devices in the input tensors" logic. However yes, we should add more support for them later.

eellison reviewed (commented):
Generally looks good... A few main comments:

- Can we merge this device/dtype analysis ? Theres a lot of duplicated code, in practice there is no scenario when who you would run one but not the other. There's a reason they are traditionally coupled. 
- I think maintaining `hasChanged` will be pretty annoying, and sometimes make this is a lot slower... I don't know if it's worth it
- A couple comments around correctness re `aten::cpu/cuda` and zero-dim tensors

Gamrix commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+    device_prop_registry_ = std::make_unique<OperatorMap<PropRule>>();
+
+    /*
+    device_prop_registry_->insert(
```
I could interactive rebase that out to merge it with the diff that uncomments and finalizes that code.

Gamrix commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+}
+
+
+bool isDeviceArgumentType(const TypePtr type) {
```
Didn't realize that that function did the right thing for Union types. (but I checked and it does).

Gamrix commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+    if (!tensor_type) {
+      continue;
+    }
+    if (device) {
```
The `if (device)` has an `else` condition.

Gamrix commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  return device.has_value() && setReturnsToDevice(n, device.value());
```
To confirm: Because Torchscript is SSA, if there is a device that is already set, it must have been set by something other than Device Propagation, right?

Is the concern that some other pass would have set an incorrect type?

eellison replied:
Yea, it's basically that the dtypes in the graph should reflect what we can currently infer. As an edge case, imagine that we had a constant tensor of dim 0 that made it so we were correctly inferring cpu as an output of an broadcast node if the other was a cpu tensor but failed if the other was a cuda tensor. if someone updated the input dtype to cuda, we shouldn't have the prior, no longer relevant dtype present on the node

Gamrix commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+  return changed;
+}
+
+bool propWithNoDevice(Node* n) {
```
As far as I am aware, no.

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,237 @@
+bool setDeviceType(Value* value, Device device, bool can_overwrite = false) {
+  auto tensor_type = value->type()->cast<TensorType>();
+  TORCH_INTERNAL_ASSERT(tensor_type, "Expecting a tensor type");
+  if (!tensor_type->device().has_value()) {
```
I reduced the locations where it is tracked, but it is needed to determine if we need to do an additional pass for analysis for loops

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+*/
+
+bool setDeviceType(Value* value, c10::optional<Device> device) {
+  auto tensor_type = value->type()->cast<TensorType>();
```
`nit` 

```
TORCH_INTERNAL_ASSERT(tensor_type, "Expecting a tensor type");
bool changed = tensor_type->device() != device;
```
-> 
```auto tensor_type = value->type()->expect<TensorType>();```

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+applies the relevant properties to the Tensor outputs
+of the Node (based on the rule itself)
+
+Returns: Bool indicating if anything was changed
```
I don't know if tracking `changed` here is worth the extra verbosity and difficulty to maintain. I think i commented this somewhere else, but I can't really think of a scenario where this would be useful in the context of dtype/device analysis. We did it in the optimization / graph manipulation passes in order to iteratively optimize.

Gamrix replied:
I have removed changed tracking for anywhere that it is not useful. Currently it is needed to know when the types in the for loop construct are consistent. (aka when loop inner start types do not change anymore). 

I have also kept it for the overall pass because that was quite easy.

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+      continue;
+    }
+
+    bool is_zerodim = tensor_type->symbolic_sizes().rank().value_or(-1) == 0;
```
`is_zerodim` - don't we want to conservatively check if it's zerodim? Here we are saying, if we don't know the rank, assume it is not zerodim

Gamrix replied:
Yes, the conservative case is that it is not zerodim, as non-zerodim tensors must have the same device, whereas zerodim CPU tensors can be used with other tensors.

eellison replied:
We're not being conservative here - we're saying that if we don't know the rank, we assume `is_zerodim = -1 ==0` which is false. The conservative assumption is that it is zerodim

eellison replied:
We're not being conservative here - 
```
@torch.jit.script
def foo(x, y):
   return x + y
```

if we set x to be cpu, but not have a rank, then we will have `is_zerodim = (-1 == 0) = False`. Then, we'll propagate the output type to be cpu because it's not zerodim. 
`foo(torch.tensor(0.3), torch.rand([2, 2]).cuda()) -> cuda tensor`. the analysis is not being conservative here.

Gamrix replied:
Incorrect. The current code would return `c10::nullopt` as the device type in the case without shape info because there are two conflicting types.

Gamrix replied:
If you want more extensive testing than the below link for zerodim, let me know what you want so that I can assure you that the code is correct.

https://github.com/pytorch/pytorch/pull/69052/files#diff-8c9c2b2ce38114d5a65e06e990ae810ea68f4dd2f54e80fc65ec7cf3d53b5c2eR105-R147

eellison replied:
> Incorrect. The current code would return c10::nullopt as the device type in the case without shape info because there are two conflicting types.

In this case we are leaving `y` not set. So we're inferring `cpu`, even though as the example shows there is a valid invocation that would produce a `cuda` tensor.

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    bool is_zerodim = tensor_type->symbolic_sizes().rank().value_or(-1) == 0;
+    bool is_cpu = tensor_type->device() && tensor_type->device()->is_cpu();
```
likewise, wouldn't we want to be conservative heree

eellison replied:
We're not being conservative here either

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+  // other type can overwrite
+
+  c10::optional<Device> device;
+  bool seen_any_device = false;
```
do we need to have a `seen_any_device` boolean ? doesn't `c10::optional<Device> device == c10::nullopt` encode that?

Gamrix replied:
No, it does not. A `c10::nullopt` Device after device propagation means that a tensor can have multiple possible device_types. 

In this function, if you have already seen an input with a `c10::nullopt` device, you cannot safely assume a different tensor's device can overwrite that multiple possible device_types. On the other hand, if you haven't seen a device yet, you can overwrite that implicit `c10::nullopt` value.

eellison replied:
> . A c10::nullopt Device after device propagation means that a tensor can have multiple possible device_types.
Why not just bail as soon as we cannot infer? What is the point of retaining this information as local state?

Gamrix replied:
You asked me to overwrite previously infered types if we infer a different type/ can't infer a type. Setting the output as `c10::nullopt` is the way to "bail". 

I don't see the advantage of erroring out over marking it as `c10::nullopt`

eellison replied:
> You asked me to overwrite previously infered types if we infer a different type/ can't infer a type. Setting the output as c10::nullopt is the way to "bail".

Yea, that's not incompatible with returning from a function when we cannot infer an output instead of adding extra state that we have to keep track of. You could factor this part out into a `c10::optional<Device>` that is returned and then set on the output. 

As a general rule if you can minimize the amount of state / control flow a reader has to keep track of in function you should

eellison replied:
Having three different pieces of local state makes this pretty hard to review and reason about - you can bail on analysis while still overwriting the output. You just factor this part of the inference out to a function that returns `c10::optional<Device>` and then set the output equal to that value. Then you can delete `seen_any_device`.

Generally, if you can minimize the amount of state and control flow in a piece of code you should do so. Also it's generally performant to exit out of a loop as soon as you know the outcome instead of continuing to iterate.

Gamrix replied:
I think the root of the issue is that you think this function should exit as soon as there is only one valid output device that we can infer. I think that is ok if we are manually whitelisting ops for Device Type Analysis (which you seem to be very explicitly against). 

The default inference I have written instead only infers if all inputs have the same device (except CPU zerodim) in order to be more conservative and checks for the case of mismatching device types, as I don't think we can by assume that we have made custom cases for all of the ops that can mix device types.

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+}
+
+bool defaultDeviceProp(Node* n) {
+  // Detecting of the op has a device object argument
```
`/s/of/if`

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+    return false;
+  }
+  auto arguments = schema->arguments();
+  auto input_vec = n->inputs();
```
unused

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+
+    switch (n->kind()) {
+      case prim::Constant:
+        // This is already been propagated by something else in freezing
```
we would like this to work on non-freezing context, could we update ?

eellison commented on torch/csrc/jit/passes/device_type_analysis.cpp:
```diff
@@ -0,0 +1,211 @@
+    }
+  }
+
+  // for efficiency
```
what does `for efficiency ` mean here ?

Gamrix replied:
Don't remember, removed.

eellison commented on torch/csrc/jit/passes/device_type_analysis.h:
```diff
@@ -0,0 +1,16 @@
+namespace jit {
+struct Graph;
+
+// Transposes the weight matrix for frozen linear modules.
```
copy-pasta

eellison reviewed (commented):
nice! any tests ?

Gamrix commented:
@eellison Tests are in diff 2 of this stack

https://github.com/pytorch/pytorch/pull/69052

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix commented:
Also, in case it wasn't clear, we have agreed that we will merge the two implementations, but this diff is already big as-is (and that is not even including the testing, which is in a different diff), and it is not critical for the current milestone. I will work on it as soon as the current stuff is landed. 

I addressed the zero-dim tensor issues. 

I have removed most of the `hasChanged` parts in the code.

eellison reviewed (approved):
Nit, I still think we could rewrite `propWithNoDevice` to be easier to follow and require less local state

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix committed:
Update on "Building Default Inference for Device Type"




[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33555885).
